src/riak_kv_qry_compiler.erl

```diff
@@ -1159,6 +1174,56 @@ modify_where_key(TupleList, Field, NewVal) ->
     {Field, FieldType, _OldVal} = lists:keyfind(Field, 1, TupleList),
     lists:keyreplace(Field, 1, TupleList, {Field, FieldType, NewVal}).
 
+resolve_expressions(Mod, Options, WhereAST) ->
+    Acc = acc, %% not used
+    riak_ql_ddl:mapfold_where_tree(
+        fun (_, Op, Acc_x) when Op == and_; Op == or_ ->
+                {ok, Acc_x};
+            (_, Filter, Acc_x) ->
+                {resolve_expressions_folder(Mod, Options, Filter), Acc_x}
+        end, Acc, WhereAST).
+
+resolve_expressions_folder(_, _, {ExpOp,{_,A},{_,B}}) when ExpOp == '+'; ExpOp == '*';
+                                                           ExpOp == '-'; ExpOp == '/' ->
+    Value =
+        case ExpOp of
+            '+' -> riak_ql_window_agg_fns:add(A,B);
+            '*' -> riak_ql_window_agg_fns:multiply(A,B);
+            '-' -> riak_ql_window_agg_fns:subtract(A,B);
+            '/' -> riak_ql_window_agg_fns:divide(A,B)
+        end,
+    {calculated, Value};
+resolve_expressions_folder(_, Options, {{sql_select_fn,'NOW'}, Args}) ->
+    ok = validate_where_fn_arity('NOW', Args),
+    {_, Now} = lists:keyfind(now_milliseconds, 1, Options),
+    {timestamp, Now};
+resolve_expressions_folder(Mod, _, {ExpOp,FieldName,{calculated,V1}}) when is_binary(FieldName) ->
+    V2 = cast_value_to_ast(Mod:get_field_type([FieldName]),V1),
+    {ExpOp,FieldName,V2};
+resolve_expressions_folder(_, _, AST) ->
+    AST.
+
+validate_where_fn_arity(FnName, Args) ->
+    ExpectedArity = riak_ql_ddl:sql_function_arity(FnName),
+    ActualArity = length(Args),
+    case ExpectedArity == length(Args) of
+        true ->
+            ok;
+        false ->
+            Msg = format_binary(
+                "Function ~p has arity of ~p but was called with ~p arguments.",
+                [FnName, ExpectedArity, ActualArity]),
+            throw({error,{invalid_query,  Msg}})
+    end.
+
+format_binary(Fmt, Args) ->
+    iolist_to_binary(io_lib:format(Fmt, Args)).
+
+cast_value_to_ast(T, V) when (T == integer orelse T == timestamp), is_integer(V) -> {T, V};
+cast_value_to_ast(T, V) when (T == integer orelse T == timestamp), is_float(V)   -> {T, erlang:round(V)};
```

Comment:
With expressions rounded to column type, will conditions like `WHERE a = 1 + 0.1` still evaluate to `true` if the column `a` is of type `sint64` and fetch the row with `a = 1`?